### DISCUSSION TITLE
`experimental.typedRoutes` with Turbopack

### DISCUSSION THREAD

**Discussion by karlhorky** [vercel/next.js#66217] [Ideas]
### Goals

1. Support parity with webpack support for `experimental.typedRoutes`

### Non-Goals

--

### Background

This is a feature request tracking issue and followup to the comments / issues talking about `experimental.typedRoutes` with Turbopack:

- https://github.com/vercel/next.js/issues/57790
  - @Audrey-Ann "experimental.typedRoutes is not supported"
- https://github.com/vercel/next.js/issues/49174#issuecomment-2088215515
  - @pawelblaszczyk5 "Any plans for experimental.typedRoutes in turbo?"
  - @timneutkens "Eventually, yes" 

### Proposal

Support `experimental.typedRoutes` with Turbopack dev server

cc @shuding

**Comment by antoinerousseau** [comment]
Hi, any news about this? That would be great to have it :)

**Comment by grmkris** [comment]
👀
<img width="809" alt="image" src="https://github.com/user-attachments/assets/089277ab-47f6-433c-b111-a9a9d4559d64">


turbo is now stable in dev, typed routes are quite nice to work with

hopefully we can get this soon?

**Comment by tjx666** [comment]
My personal script has better ux.

I use following code to generate env types:


```typescript
/**
 * Auto generate environment.d.ts by .env.local
 */
import '@/scripts/checkCwd';

import { readFile, writeFile } from 'fs/promises';
import { resolve } from 'path';

import { $ } from 'execa';

async function parseEnvFile(filePath: string): Promise<Array<[string, string, string | null]>> {
    const content = await readFile(filePath, 'utf8');
    const lines = content.split('\n');
    const result: Array<[string, string, string | null]> = [];
    let lastComment: string | null = null;

    for (const line of lines) {
        const trimmedLine = line.trim();
        if (trimmedLine.startsWith('#')) {
            lastComment = trimmedLine.slice(1).trim();
        } else if (trimmedLine.includes('=')) {
            const [key, value] = trimmedLine.split('=').map((part) => part.trim());
            result.push([key, value.replaceAll(/^['"]|['"]$/g, ''), lastComment]);
            lastComment = null;
        }
    }

    return result;
}

const envEntries = await parseEnvFile(resolve('.env.local'));

const indent = ' '.repeat(4);
const envs = envEntries
    .map(([key, value, comment]) => {
        const commentBlockLines = [
            '/**',
            comment ? ` * ${comment}` : '',
            comment ? ' *' : '',
            ' * @example',
            ` *     '${value}'`,
            ' */',
        ];
        const commentBlock = commentBlockLines
            .filter(Boolean)
            .map((line) => indent.repeat(3) + line)
            .join('\n');
        return `${commentBlock}\n${indent.repeat(3)}${key}: string;`;
    })
    .join('\n\n');

const content = `/**
 * !: DON'T EDIT THIS FILE MANUALLY
 *
 * Auto generate by scripts/updateEnvDts.ts
 */

import Next from 'next'

declare global {
    namespace NodeJS {
        interface ProcessEnv {
${envs}
        }
    }
}
`;

await writeFile(resolve('typings/environment.d.ts'), content);
await $`prettier --write typings/environment.d.ts`;
```

result:

```typescript
/**
 * !: DON'T EDIT THIS FILE MANUALLY
 *
 * Auto generate by scripts/updateEnvDts.ts
 */

import Next from 'next';

declare global {
    namespace NodeJS {
        interface ProcessEnv {
            /**
             * Clerk Auth
             *
             * @example
             *     '/sign-in';
             */
            NEXT_PUBLIC_CLERK_SIGN_IN_URL: string;

            /**
             * @example
             *     '/sign-up';
             */
            NEXT_PUBLIC_CLERK_SIGN_UP_URL: string;

            /**
             * @example
             *     'pk_************************************Q';
             */
            NEXT_PUBLIC_CLERK_PUBLISHABLE_KEY: string;

            /**
             * @example
             *     'sk_************************************i';
             */
            CLERK_SECRET_KEY: string;

            /**
             * FAL API
             *
             * @example
             *     'ab_************************************8';
             */
            FAL_KEY: string;

            /**
             * Neon Postgres
             *
             * @example
             *     'pg_************************************e';
             */
            DATABASE_URL: string;

            /**
             * https://dash.cloudflare.com/c9a6a0fa818553a00efc0d8e5/r2/api-tokens
             *
             * @example
             *     'cf_************************************j';
             */
            R2_TOKEN: string;

            /**
             * @example
             *     'r2_************************************0';
             */
            R2_ACCESS_KEY_ID: string;

            /**
             * @example
             *     'sk_************************************f';
             */
            R2_SECRET_ACCESS_KEY: string;
        }
    }
}

```

**Comment by johanneskares** [comment]
This is my very clunky workaround. I added this to scripts in `package.json`

```json
{
    "dev": "bun run delete-typed-routes && next dev --turbopack",
    "build": "bun run add-typed-routes && bun run next build --no-lint",
    "add-typed-routes": "grep -qxF '        typedRoutes: true,' next.config.js || (sed -i.bak '5s/^/        typedRoutes: true,\n/' next.config.js && rm -f next.config.js.bak)",
    "delete-typed-routes": "sed -i.bak '/typedRoutes: true/d' next.config.js && rm -f next.config.js.bak"
}
```

It basically disables type checking for the dev build, but I still get errors during build phase.

**Reply by karlhorky** [reply]
Ah yeah, something like that is what I was thinking over here:

- https://twitter.com/karlhorky/status/1846506414233067993

Indeed a bit hacky, but maybe a good option until Turbopack support.

**Reply by javascripter** [reply]
You can use environment variables instead without modifying config files.

next.config.ts
```typescript 
import type { NextConfig } from 'next'

const nextConfig: NextConfig = {
  experimental: {
    typedRoutes: ['1', 'true'].includes(process.env.TYPED_ROUTES ?? ''),
  },
}

export default nextConfig
```

Then,
```bash
TYPED_ROUTES=1 bun run build
```

```bash
bun run dev
```

**Comment by HashimMufti** [comment]
Gentle bump on this, would be great if we didn't have to hack around it!

**Comment by karlhorky** [comment]
[`next@15.5.0`](https://github.com/vercel/next.js/releases/tag/v15.5.0) includes `typedRoutes` for Turbopack (thanks @bgub and reviewers! 🎉):

- Blog post: https://nextjs.org/blog/next-15-5
- Full release notes: https://github.com/vercel/next.js/releases/tag/v15.5.0

---

Looks like `experimental.typedRoutes` is coming to Turbopack in these PRs, thanks to @bgub 🔥 🚀

- https://github.com/vercel/next.js/pull/81396
- https://github.com/vercel/next.js/pull/81400
- https://github.com/vercel/next.js/pull/81528

<img width="1922" height="608" alt="Screenshot 2025-07-17 at 10 50 12" src="https://github.com/user-attachments/assets/77a4c5ff-8914-4f69-845c-d58492851686" />

**Reply by HashimMufti** [reply]
Great news, thanks for sharing!

**Reply by lanzclintonv** [reply]
As of `15.4.7` we still don't have typedRoutes in turbo

**Reply by karlhorky** [reply]
The latest patch releases don't include all of the changes in canary - Next.js patch releases don't do that anymore. 

But it looks like a new minor `next@15.5.0` was just released, with `typedRoutes` for Turbopack (thanks @bgub and @lukesandberg 🎉):

- https://github.com/vercel/next.js/releases/tag/v15.5.0
- https://nextjs.org/blog/next-15-5

I updated my original post above 👍

**Reply by lanzclintonv** [reply]
Yeah, it dropped literally today. I can confirm that it's working client side. 

Unfortunately, `typedRoutes` is still unavailable in `redirect()`

**Reply by bgub** [reply]
@lanzclintonv could you share more details? If there's an easy solution, I'll see what I can do. We do handle some cases of redirects...

**Reply by lanzclintonv** [reply]
@bgub sure. The type safe navigation doesn't work with redirect from `next/navigation`. The href in Link and router.push gives type safety.

I checked the repo and it seems [redirect is fixed to the string type](https://github.com/vercel/next.js/blob/canary/packages/next/src/client/components/redirect.ts#L40). 

I'm not familiar with Next.js repo and haven't had the time yet. I guess the solution would be replacing the string type with the same type of href in Link and router.push() params. I'm in no way an expert on this so take it with a grain of salt.

**Reply by bgub** [reply]
@lanzclintonv just added support!

https://github.com/vercel/next.js/pull/82860

**Reply by lanzclintonv** [reply]
@bgub can confirm that `redirect` and `permanentRedirect` now has typedRoutes as of `15.5.1`. Thanks for fixing this!
